### PR TITLE
19387 update endpoints to work with alt name

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#19387

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

**Comment by argush3** [comment]
@kzdev420 have you tested at least the getters on these endpoints locally to see if things work for the endpoints you updated?

**Comment by kzdev420** [comment]
> @kzdev420 have you tested at least the getters on these endpoints locally to see if things work for the endpoints you updated?

Sorry, I didn't test, because I had some issue to install the env.

**Comment by argush3** [comment]
> > @kzdev420 have you tested at least the getters on these endpoints locally to see if things work for the endpoints you updated?
> 
> Sorry, I didn't test, because I had some issue to install the env.

I think you should get the api running locally with the db dump I shared so we have an idea if what you've implemented actually works.  

We will have future tickets to fix all the tests but I would still like to see that what you have implemented actually works to an extent.

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/request_tracker.py:38]: This won't work as is when the business is a SP.  

When the business is a SP, ` business_service.fetch_business(identifier)` will return an instance of `alternate_name` which means that `RequestTracker.find_by` will return an incorrect request tracker instance or return nothing.  `RequestTracker.find_by` needs to be searching by `alternate_name_id` in this instance.

You should be doing something similar to this:

https://github.com/bcgov/lear/blob/feature-legal-name/legal-api/src/legal_api/models/filing.py#L865-L873

I suspect there are a bunch of other models you'll need to do this for.  Off the top of my head, I know that `documents`, `filings`, `comments`, `request_tracker` and `offices` are tied to `alternate_names` for SPs.

File: legal-api/src/legal_api/models/legal_entity.py
Comment (argush3) [inline legal-api/src/legal_api/models/legal_entity.py:594]: I don't think we want to update `LegalEntity.find_by_internal_id` like this.  You need have access to the right business(AlternateName or LegalEntity) before you call the `find_by_internal_id` function

File: legal-api/src/legal_api/resources/v2/business/business_address.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_address.py:91]: you need to check this.  `Address` has `alternate_name_id` too.

File: legal-api/src/legal_api/resources/v2/business/business_address.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_address.py:87]: think this should be `business_id_attr = AlternateName.id`

File: legal-api/src/legal_api/resources/v2/business/business_comments.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_comments.py:54]: I believe @leodube-aot is going to be adding a function to `business_service` to provide this functionality in a reusable function.  Once he has that code merged, you can rebase and use that new function here.

File: legal-api/src/legal_api/resources/v2/business/business_documents.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_documents.py:111]: The business passed into `_get_coop_documents_and_info` will always be a `LegalEntity` instance as this function is specific to a COOP.  i.e. you can just use business.id

File: legal-api/src/legal_api/resources/v2/business/business_filings/business_documents.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_filings/business_documents.py:72]: as mentioned previously, there will be a function for this in `business_service` soon.

File: legal-api/src/legal_api/resources/v2/business/business_parties.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_parties.py:48]: This code doesn't work for SPs.  SPs don't use `entity_roles` table except for the completing party which is not relevant to this function.  

There are 3 scenarios that you will need to take into account - SP individual, SP LEAR DBA, SP COLIN DBA.

You will need to write code to still return what the /parties endpoint previously returned for proprietors.

All 3 of these scenarios will be a bit different both in the owning record as well as the addresses to use.

SP individual - use person LE record tied to alternate name record.  also use `mailing_address_id` and `delivery_address_id` for person LE record.

SP LEAR DBA - use owning business LE record tied to alternate name record.  Use `mailing_address_id` and `delivery_address_id` from the alternate name record.

SP COLIN DBA - use owning business colin entity record tied to alternate name record.  Use `mailing_address_id` and `delivery_address_id` from the colin entity record.

File: legal-api/src/legal_api/resources/v2/business/business_resolutions.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_resolutions.py:54]: this function will only ever be called for non-firms.  so you can always use `business.id`.  don't need this line.

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:75]: not needed.  just use business.id

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:78]: you'll need to update `Filing.get_filings_by_status` to work with a business(alternate name or legal entity) instead of business_id.

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:123]: as per comment above, needs to work with business instance instead of business.id

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:108]: remove, can use business.id

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:155]: is this used?

File: legal-api/src/legal_api/resources/v2/business/filing_comments.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/filing_comments.py:144]: use function in `business_service`

File: legal-api/src/legal_api/resources/v2/business/filing_comments.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/filing_comments.py:154]: you can just use `filing_business_id = filing.legal_entity_id or filing.alternate_name_id`

File: legal-api/src/legal_api/models/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/models/request_tracker.py:93]: can remove.  just use `business`

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/request_tracker.py:40]: please update to not pass `business_id` into `RequestTracker.find_by`

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/request_tracker.py:73]: this will need to be updated to support SPs.  SPs will use `alternate_name_id` in `RequestTracker` model instead of `legal_entity_id`

File: legal-api/src/legal_api/resources/v2/business/business_parties.py
Comment (argush3) [reply legal-api/src/legal_api/resources/v2/business/business_parties.py:48]: Actually, don't worry about the address bits here for now.  I forgot we have a ticket(#19667) specifically to address work.

Just update things such that a proprietor is at least returned properly for SPs when the GET parties endpoint is called.

File: legal-api/src/legal_api/resources/v2/business/business_address.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_address.py:91]: Updated

File: legal-api/src/legal_api/resources/v2/business/business_address.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_address.py:87]: Updated

File: legal-api/src/legal_api/resources/v2/business/business_documents.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_documents.py:111]: updated

File: legal-api/src/legal_api/resources/v2/business/business_resolutions.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_resolutions.py:54]: updated

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py:78]: updated

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py:123]: updated

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py:108]: updated

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py:155]: used

File: legal-api/src/legal_api/resources/v2/business/filing_comments.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/filing_comments.py:154]: Updated

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/request_tracker.py:73]: updated

File: legal-api/src/legal_api/resources/v2/business/business_comments.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_comments.py:54]: updated

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py]: updated

File: legal-api/src/legal_api/resources/v2/business/filing_comments.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/filing_comments.py:144]: updated

File: legal-api/src/legal_api/models/request_tracker.py
Comment (kzdev420) [reply legal-api/src/legal_api/models/request_tracker.py]: updated

**Review by JazzarKarim — APPROVED** [review]
LGTM but please wait for another approval before merging 👍

File: legal-api/src/legal_api/models/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/models/request_tracker.py:99]: can we remove `legal_entity_id` from the parameter list?

File: legal-api/src/legal_api/models/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/models/request_tracker.py:108]: update to use `business.id`

File: legal-api/src/legal_api/resources/v2/business/business_documents.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_documents.py:87]: i don't think you need L86.  you should be able to just use `if business.entity_type == LegalEntity.EntityTypes.COOP.value:`

File: legal-api/src/legal_api/models/filing.py
Comment (argush3) [inline legal-api/src/legal_api/models/filing.py:736]: can we remove `business_id`

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_tasks.py:76]: remove `business.id` after you update `Filing.get_filings_by_status`

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/request_tracker.py:39]: remove `business.id` param after removing this param from `RequestTracker.find_by`

File: legal-api/src/legal_api/resources/v2/business/business_parties.py
Comment (argush3) [reply legal-api/src/legal_api/resources/v2/business/business_parties.py:48]: so this code is still a problem.  

Try the following endpoint against in dev and you should see what you should expect to get back.

``` json
GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/FM1028594/parties
{
  "parties": [
    {
      "deliveryAddress": {
        "addressCity": "Langley",
        "addressCountry": "CA",
        "addressRegion": "BC",
        "deliveryInstructions": null,
        "id": 2649561,
        "postalCode": "V2Z 2X9",
        "streetAddress": "23434 6 Ave",
        "streetAddressAdditional": null
      },
      "mailingAddress": {
        "addressCity": "Langley",
        "addressCountry": "CA",
        "addressRegion": "BC",
        "deliveryInstructions": null,
        "id": 2649562,
        "postalCode": "V2Z 2X9",
        "streetAddress": "23434 6 Ave",
        "streetAddressAdditional": null
      },
      "officer": {
        "email": "someemail..com",
        "firstName": "JANE",
        "id": 568475,
        "lastName": "DOE",
        "partyType": "person"
      },
      "roles": [
        {
          "appointmentDate": "2023-02-02",
          "cessationDate": null,
          "roleType": "Proprietor"
        }
      ]
    }
  ]
}
```

If you test locally, I don't think it won't return the same response, as the proprietor is not linked via the entity_roles table.  

As I had mentioned in my first comment, you will need to handle the 3 SP scenarios.  Don't worry about the details around which address to return for the proprietor for now.

File: legal-api/src/legal_api/models/request_tracker.py
Comment (kzdev420) [reply legal-api/src/legal_api/models/request_tracker.py:99]: Yes, I can do it.
But, `queue-service/entity-bn` used this func
https://github.com/kzdev420/lear/blob/a1b1c98c8ceb8d5516a5e3b385ab8b9d07817a91/queue_services/entity-bn/src/entity_bn/bn_processors/admin.py#L41-L47

If I updated the request_tracker model, should I update the queue-service too?

Comment (argush3) [reply legal-api/src/legal_api/models/request_tracker.py:99]: don't worry about other services for now.  we have tickets for that work.

File: legal-api/src/legal_api/models/request_tracker.py
Comment (kzdev420) [reply legal-api/src/legal_api/models/request_tracker.py:108]: Same as above comment.

File: legal-api/src/legal_api/resources/v2/business/business_documents.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_documents.py:87]: fixed

File: legal-api/src/legal_api/resources/v2/business/business_tasks.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_tasks.py:76]: fixed

File: legal-api/src/legal_api/resources/v2/request_tracker.py
Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/request_tracker.py:39]: fixed

File: legal-api/src/legal_api/resources/v2/business/business_parties.py
Comment (argush3) [reply legal-api/src/legal_api/resources/v2/business/business_parties.py:48]: As per our conversation, you can fix this in your next PR.

File: legal-api/src/legal_api/resources/v2/business/business_documents.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business_documents.py:86]: L86 should be `if business.entity_type == LegalEntity.EntityTypes.COOP.value:`

Comment (kzdev420) [reply legal-api/src/legal_api/resources/v2/business/business_documents.py]: fixed

**Review by argush3 — APPROVED** [review]

**Pull request merged**
